// @nuxt/devtools/dist/runtime/plugins/view/client.mjs
import { computed, createApp, h, markRaw, nextTick, ref, shallowReactive, shallowRef, watch } from "vue";
import { createHooks } from "hookable";
import Main from "./Main.vue";
import { popupWindow, state } from "./state.mjs";
import { useAppConfig } from "#imports";
const clientRef = shallowRef();
export { clientRef as client };
export async function setupDevToolsClient({
  nuxt,
  clientHooks,
  timeMetric
}) {
  const isInspecting = ref(false);
  const colorMode = useClientColorMode();
  const client = shallowReactive({
    nuxt: markRaw(nuxt),
    appConfig: useAppConfig(),
    hooks: createHooks(),
    getClientHooksMetrics: () => Object.values(clientHooks),
    getClientPluginMetrics: () => {
      return window.__NUXT_DEVTOOLS_PLUGINS_METRIC__ || [];
    },
    loadingTimeMetrics: timeMetric,
    reloadPage() {
      location.reload();
    },
    toggle() {
      if (state.value.open)
        client.close();
      else
        client.open();
    },
    close() {
      if (!state.value.open)
        return;
      state.value.open = false;
      if (popupWindow.value) {
        try {
          popupWindow.value.close();
        } catch (e) {
        }
        popupWindow.value = null;
      }
    },
    open() {
      if (state.value.open)
        return;
      state.value.open = true;
      nextTick(() => {
        client.updateClient();
      });
    },
    inspector: getInspectorInstance(),
    colorMode,
    getIframe,
    updateClient
  });
  let iframe;
  function updateClient() {
    if (!client.inspector)
      client.inspector = getInspectorInstance();
    try {
      iframe?.contentWindow?.__NUXT_DEVTOOLS_VIEW__?.setClient(client);
    } catch (e) {
    }
    return client;
  }
  function getIframe() {
    if (!iframe) {
      const CLIENT_PATH = "/__nuxt_devtools__/client";
      const initialUrl = CLIENT_PATH + state.value.route;
      try {
        iframe = document.createElement("iframe");
        iframe.id = "nuxt-devtools-iframe";
        iframe.src = initialUrl;
        iframe.onload = async () => {
          await waitForClientInjection();
          client.updateClient();
        };
      } catch (e) {
        console.error(e);
      }
    }
    return iframe;
  }
  function waitForClientInjection(retry = 10, timeout = 200) {
    const test = () => {
      try {
        return !!iframe?.contentWindow?.__NUXT_DEVTOOLS_VIEW__;
      } catch (e) {
      }
      return false;
    };
    if (test())
      return;
    return new Promise((resolve, reject) => {
      const interval = setInterval(() => {
        if (test()) {
          clearInterval(interval);
          resolve();
        } else if (retry-- <= 0) {
          clearInterval(interval);
          reject("Nuxt Devtools client injection failed");
        }
      }, timeout);
    });
  }
  function enableComponentInspector() {
    window.__VUE_INSPECTOR__?.enable();
    isInspecting.value = true;
  }
  function disableComponentInspector() {
    if (!window.__VUE_INSPECTOR__?.enabled)
      return;
    window.__VUE_INSPECTOR__?.disable();
    client?.hooks.callHook("host:inspector:close");
    isInspecting.value = false;
  }
  function getInspectorInstance() {
    const componentInspector = window.__VUE_INSPECTOR__;
    if (componentInspector) {
      componentInspector.openInEditor = async (baseUrl, file, line, column) => {
        disableComponentInspector();
        await client.hooks.callHook("host:inspector:click", baseUrl, file, line, column);
      };
      componentInspector.onUpdated = () => {
        client.hooks.callHook("host:inspector:update", {
          ...componentInspector.linkParams,
          ...componentInspector.position
        });
      };
    }
    return markRaw({
      isEnabled: isInspecting,
      enable: enableComponentInspector,
      disable: disableComponentInspector,
      toggle: () => {
        if (window.__VUE_INSPECTOR__?.enabled)
          disableComponentInspector();
        else
          enableComponentInspector();
      },
      instance: componentInspector
    });
  }
  function refreshReactivity() {
    client.hooks.callHook("host:update:reactivity");
  }
  watch(() => [
    client.nuxt.payload,
    client.colorMode.value,
    client.loadingTimeMetrics
  ], () => {
    refreshReactivity();
  }, { deep: true });
  client.nuxt.vueApp.config.globalProperties?.$router?.afterEach(() => {
    refreshReactivity();
  });
  client.nuxt.hook("app:mounted", () => {
    refreshReactivity();
  });
  client.hooks.hook("devtools:navigate", (path) => {
    state.value.route = path;
  });
  clientRef.value = client;
  const documentPictureInPicture = window.documentPictureInPicture;
  if (documentPictureInPicture?.requestWindow) {
    client.popup = async () => {
      const iframe2 = getIframe();
      if (!iframe2)
        return;
      const pip = popupWindow.value = await documentPictureInPicture.requestWindow({
        width: Math.round(window.innerWidth * state.value.width / 100),
        height: Math.round(window.innerHeight * state.value.height / 100)
      });
      const style = pip.document.createElement("style");
      style.innerHTML = `
        body {
          margin: 0;
          padding: 0;
        }
        iframe {
          width: 100vw;
          height: 100vh;
          border: none;
          outline: none;
        }
      `;
      pip.__NUXT_DEVTOOLS_DISABLE__ = true;
      pip.__NUXT_DEVTOOLS_IS_POPUP__ = true;
      pip.document.title = "Nuxt DevTools";
      pip.document.head.appendChild(style);
      pip.document.body.appendChild(iframe2);
      pip.addEventListener("resize", () => {
        state.value.width = Math.round(pip.innerWidth / window.innerWidth * 100);
        state.value.height = Math.round(pip.innerHeight / window.innerHeight * 100);
      });
      pip.addEventListener("pagehide", () => {
        popupWindow.value = null;
        pip.close();
      });
    };
  }
  client.updateClient();
  const holder = document.createElement("div");
  holder.id = "nuxt-devtools-container";
  holder.setAttribute("data-v-inspector-ignore", "true");
  document.body.appendChild(holder);
  addEventListener("keydown", (e) => {
    if (e.code === "KeyD" && e.altKey && e.shiftKey)
      client.close();
  });
  const app = createApp({
    render: () => h(Main, { client }),
    devtools: {
      hide: true
    }
  });
  app.mount(holder);
}
export function useClientColorMode() {
  const explicitColor = ref();
  const systemColor = ref();
  const elements = [
    document.documentElement,
    document.body
  ];
  const ob = new MutationObserver(getExplicitColor);
  elements.forEach((el) => {
    ob.observe(el, {
      attributes: true,
      attributeFilter: ["class"]
    });
  });
  const preferDarkQuery = window.matchMedia("(prefers-color-scheme: dark)");
  const preferLightQuery = window.matchMedia("(prefers-color-scheme: light)");
  preferDarkQuery.addEventListener("change", getSystemColor);
  preferLightQuery.addEventListener("change", getSystemColor);
  function getExplicitColor() {
    let color;
    for (const el of elements) {
      if (el.classList.contains("dark")) {
        color = "dark";
        break;
      }
      if (el.classList.contains("light")) {
        color = "light";
        break;
      }
    }
    explicitColor.value = color;
  }
  function getSystemColor() {
    if (preferDarkQuery.matches)
      systemColor.value = "dark";
    else if (preferLightQuery.matches)
      systemColor.value = "light";
    else
      systemColor.value = void 0;
  }
  getExplicitColor();
  getSystemColor();
  return computed(() => explicitColor.value || systemColor.value || "light");
}
